backend/app.js

```diff
@@ -107,55 +107,25 @@ passport.use(new LocalStrategy(
-          // in the backend server, probably to make sure that no tampering took place.
-          // 
-          // todo: this aminoMsg prep and call to makeSignDoc would be shared code between
-          // the front-end and the back-end.
-          console.log("nonce before aminoMsg:", user.nonce);
-          const aminoMsg = {
-            type: 'cosmos-sdk/TextProposal',
-            value: {
-              title: "Regen Network Login Text Proposal",
-              description: 'This is a transaction that allows Regen Network to authenticate you with our application.',
-              nonce: user.nonce,
-              // proposer: address,
-              // initial_deposit: [{ denom: 'stake', amount: '0' }]
-            }
-          };
-          // generate a new nonce for the user to invalidate the current
-          // signature...
-          user.nonce = genNonce();
-          const fee = {
-            gas: '1',
-            amount: [
-              { denom: 'regen', amount: '0' }
-            ]
-          };
-          const generatedSignDoc = amino.makeSignDoc([aminoMsg], fee, 'regen-1', 'Regen Network Login Memo', '0', '0');
-          // step 3: make sure the incoming signDoc, and the one that we created above are
-          // a match. if not, the user should not be authenticated.
-          if (amino.serializeSignDoc(signed).toString() == amino.serializeSignDoc(generatedSignDoc).toString()) {
-            // step 4: given the decoded pubkey and signature from the incoming signDoc, verify that
-            // the generated signDoc from above was indeed signed by originating key.
-            const { pubkey, signature: decodedSig } = amino.decodeSignature(signature);
-            const secpSignature = crypto.Secp256k1Signature.fromFixedLength(decodedSig);
-            const messageHash = new crypto.Sha256(amino.serializeSignDoc(generatedSignDoc)).digest();
-            const isValid = await crypto.Secp256k1.verifySignature(secpSignature, messageHash, pubkey);
-            if (isValid) {
-              return done(null, fetchUserById(user.id));
-            } else {
-              return done(null, false);
-            }
+        if (address === user.address) {
+          const { pubkey: decodedPubKey, signature: decodedSignature } = amino.decodeSignature(signature);
+          const data = JSON.stringify({
+            title: 'Regen Network Login',
+            description: 'This is a transaction that allows Regen Network to authenticate you with our application.',
+            nonce: user.nonce
+          });
+          // https://github.com/chainapsis/keplr-wallet/blob/master/packages/cosmos/src/adr-36/amino.ts
+          const verified = verifyADR36Amino("regen", address, data, decodedPubKey, decodedSignature);
+          if (verified) {
+            // generate a new nonce for the user to invalidate the current
+            // signature...
+            user.nonce = genNonce();
```

Comment:
I have a question on where this should go. Maybe it should go outside the if block?

The scenario is: frontend user gets a nonce from backend, but signs incorrectly (for whatever reason). Should the next try sign the same nonce, or a different one? It feels safer to generate a new nonce even on failed attempts.

Reply:
We can do that for good measure. Initially I did have it as that but then I had a thought that it might not be necessary. But in agreement with you.